### ISSUE TITLE
Suspected presolver bug in fzn-cp-sat 9.10.4067

### ISSUE THREAD

**Issue by matsc-at-sics-se** [google/or-tools#4392]
I have a model that other solvers solve fine, but CP-SAT says "Unsat after presolving", pointing at an element constraint.
I tried with `--nopresolve`, but that flag apparently had no effect.

I can't share the model yet, maybe later.

Is there any known issue with the presolver?  Is there a better way to switch it off? Is there a way to get more details, one of the "dump" options maybe?

**Comment by lperron** [comment]
cat you try with 9.11 ? or send me the mzn/dzn of fzn file ?

**Comment by CervEdin** [comment]
I'm working with @matsc-at-sics-se on this issue

I will try and see with 9.11 and maybe some other versions.
Also checking how we can share a mzn/dzn or fzn file.

Meanwhile, I've tried turning off pre-solve.
There's a parameter in the sat proto
https://github.com/google/or-tools/blob/ed94162b910fa58896db99191378d3b71a5313af/ortools/sat/sat_parameters.proto#L440C1-L441C57
but it doesn't seem to turn off presolving, at least not the way I expect.

So running
```sh
minizinc -v --solver or-tools --params cp_model_presolve:false,keep_all_feasible_solutions_in_presolve:false bug.fzn
```
still appears to pre-solve
```
%% File /tmp/mznfileb1F4l6.fzn parsed in 0 ms
%% Presolve model
%%   - rule 'array_var_int_element: simplify using affine mapping.' was applied 3 times
%%   - rule 'int_lin_eq: store affine mapping' was applied 3 times
%%   - done in 0 ms
%% Model mznfileb1F4l6
%%   - array_int_element: 3
%%   - array_var_int_element: 3
%%   - ortools_table_int: 3
%%   - Satisfaction problem
%% The number of search workers, is not specified. For better performances, please set the number of workers to 8, 16, or more depending on the number of cores of your computer.
%% Starting CP-SAT solver v9.10.4067
%% Parameters: log_search_progress: true search_branching: FIXED_SEARCH cp_model_presolve: false keep_all_feasible_solutions_in_presolve: false log_to_stdout: false max_domain_size_when_encoding_eq_neq_constraints: 32 num_workers: 1
%% Initial satisfaction model 'mznfileb1F4l6': (model_fingerprint: 0x1398b3a42bb8c4d9)
%% #Variables: 24
%%   - 3 in [1,2]
%%   - 3 in [1,3]
%%   - 3 in [344,480]
%%   - 15 constants in {0,7,9,10,79,102,106,167,344,480} 
%% #kElement: 6
%% #kTable: 3
%% Starting presolve at 0.00s
%% INFEASIBLE: ''
%% UNSAT after expansion of table { vars: 9 vars: 12 vars: 13 values: 480 values: 9 values: 7 values: 480 values: 9 values: 10 values: 480 values: 106 values: 7 values: 480 values: 106 values: 10 }
%% Presolve summary:
%%   - 0 affine relations were detected.
%%   - rule 'element: expanded value element' was applied 2 times.
%%   - rule 'element: reduced index domain' was applied 1 time.
%%   - rule 'element: reduced target domain' was applied 1 time.
%%   - rule 'table: empty' was applied 1 time.
%%   - rule 'variables: add encoding constraint' was applied 2 times.
%% Problem closed by presolve.
%% CpSolverResponse summary:
%% status: INFEASIBLE
%% objective: NA
%% best_bound: NA
%% integers: 0
%% booleans: 0
%% conflicts: 0
%% branches: 0
%% propagations: 0
%% integer_propagations: 0
%% restarts: 0
%% lp_iterations: 0
%% walltime: 0.000473515
%% usertime: 0.000473561
%% deterministic_time: 0
%% gap_integral: 0
%% 
=====UNSATISFIABLE=====
```

@lperron am I using the parameters correctly?

**Comment by lperron** [comment]
expansion must always happen, even if the presolve is off.

CP_SAT has no implementation of the table constraint (among others).
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mer. 2 oct. 2024 à 09:01, Erik Cervin Edin ***@***.***> a
écrit :

**Comment by lperron** [comment]
can you send me bug.fzn ?
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Comment by CervEdin** [comment]
I tried with minizinc 2.8.6 and OR-Tools 9.11 and it's the same result.

This is the compiled flatzinc file:
```
predicate ortools_table_int(array [int] of var int: x,array [int] of int: t);
array [1..2] of int: X_INTRODUCED_31_ = [480,344];
array [1..12] of int: X_INTRODUCED_33_ = [480,9,7,480,9,10,480,106,7,480,106,10];
var 1..2: X_INTRODUCED_16_;
var 1..2: X_INTRODUCED_17_;
var 1..2: X_INTRODUCED_18_;
var 167..167: X_INTRODUCED_20_;
var 102..102: X_INTRODUCED_21_;
var 79..79: X_INTRODUCED_22_;
var 1..3: X_INTRODUCED_23_;
var 1..3: X_INTRODUCED_24_;
var 1..3: X_INTRODUCED_25_;
var 2..3: X_INTRODUCED_28_ ::var_is_introduced :: is_defined_var;
var 344..480: X_INTRODUCED_30_ ::var_is_introduced :: is_defined_var;
var 2..3: X_INTRODUCED_35_ ::var_is_introduced :: is_defined_var;
var 344..480: X_INTRODUCED_36_ ::var_is_introduced :: is_defined_var;
var 2..3: X_INTRODUCED_39_ ::var_is_introduced :: is_defined_var;
var 344..480: X_INTRODUCED_40_ ::var_is_introduced :: is_defined_var;
array [1..3] of var int: X_INTRODUCED_32_ ::var_is_introduced  = [X_INTRODUCED_30_,9,7];
array [1..3] of var int: X_INTRODUCED_37_ ::var_is_introduced  = [X_INTRODUCED_36_,106,10];
array [1..3] of var int: X_INTRODUCED_41_ ::var_is_introduced  = [X_INTRODUCED_40_,106,7];
constraint array_var_int_element(X_INTRODUCED_28_,[167,102,79],102);
constraint array_int_element(X_INTRODUCED_16_,X_INTRODUCED_31_,X_INTRODUCED_30_):: defines_var(X_INTRODUCED_30_);
constraint ortools_table_int(X_INTRODUCED_32_,X_INTRODUCED_33_);
constraint array_var_int_element(X_INTRODUCED_35_,[167,102,79],102);
constraint array_int_element(X_INTRODUCED_17_,X_INTRODUCED_31_,X_INTRODUCED_36_):: defines_var(X_INTRODUCED_36_);
constraint ortools_table_int(X_INTRODUCED_37_,X_INTRODUCED_33_);
constraint array_var_int_element(X_INTRODUCED_39_,[167,102,79],102);
constraint array_int_element(X_INTRODUCED_18_,X_INTRODUCED_31_,X_INTRODUCED_40_):: defines_var(X_INTRODUCED_40_);
constraint ortools_table_int(X_INTRODUCED_41_,X_INTRODUCED_33_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_16_,X_INTRODUCED_28_],-1):: domain:: defines_var(X_INTRODUCED_28_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_17_,X_INTRODUCED_35_],-1):: domain:: defines_var(X_INTRODUCED_35_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_18_,X_INTRODUCED_39_],-1):: domain:: defines_var(X_INTRODUCED_39_);
solve  satisfy;
```

I did my best at reducing the model and I ended up with this:
```
include "table.mzn";

enum B = { B1, B2 };
enum F = { F1, F2 };
enum R = { R1, R2 };
enum T = { T1, T2, T3 };

constraint forall (i in 1..3)
( let {
  var index_set(arr_var1): r  =  arr_var2[i],
  par F: foo                  =  [ F1, F2, F2 ][i],
  par B: bar                  =  [ B1, B2, B1 ][i],
  array[int] of int: arr_par  =  [ R1, R2 ],
} in
  arr_var1[r] == T3 /\
  table(
    [ arr_par[r], foo, bar ],
    [| R1, F1, B1
     | R1, F2, B1
     | R1, F2, B2 |]
  )
);

array[int] of var int: arr_var1 = array1d(0..2, [T1] ++ [ rt(r) | r in 1..2 ]);
array[1..3] of var index_set(arr_var1): arr_var2;

function var int: rt(var int: r) = [ T3, T2, T1 ][r];
```
which flattens down into
```
predicate ortools_table_int(array [int] of var int: x,array [int] of int: t);
array [1..2] of int: X_INTRODUCED_69_ = [1,2];
array [1..9] of int: X_INTRODUCED_77_ = [1,1,1,1,2,1,1,2,2];
var 1..2: X_INTRODUCED_66_;
var 1..2: X_INTRODUCED_67_;
var 1..2: X_INTRODUCED_68_;
var 1..3: X_INTRODUCED_71_ ::var_is_introduced :: is_defined_var;
var 1..2: X_INTRODUCED_73_ ::var_is_introduced :: is_defined_var;
var 1..3: X_INTRODUCED_79_ ::var_is_introduced :: is_defined_var;
var 1..2: X_INTRODUCED_80_ ::var_is_introduced :: is_defined_var;
var 1..3: X_INTRODUCED_85_ ::var_is_introduced :: is_defined_var;
var 1..2: X_INTRODUCED_86_ ::var_is_introduced :: is_defined_var;
array [1..3] of var int: arr_var2:: output_array([1..3]) = [X_INTRODUCED_66_,X_INTRODUCED_67_,X_INTRODUCED_68_];
array [1..3] of var int: X_INTRODUCED_76_ ::var_is_introduced  = [X_INTRODUCED_73_,1,1];
array [1..3] of var int: X_INTRODUCED_83_ ::var_is_introduced  = [X_INTRODUCED_80_,2,2];
array [1..3] of var int: X_INTRODUCED_89_ ::var_is_introduced  = [X_INTRODUCED_86_,2,1];
constraint array_var_int_element(X_INTRODUCED_71_,[1,3,2],3);
constraint array_int_element(X_INTRODUCED_66_,X_INTRODUCED_69_,X_INTRODUCED_73_):: defines_var(X_INTRODUCED_73_);
constraint ortools_table_int(X_INTRODUCED_76_,X_INTRODUCED_77_);
constraint array_var_int_element(X_INTRODUCED_79_,[1,3,2],3);
constraint array_int_element(X_INTRODUCED_67_,X_INTRODUCED_69_,X_INTRODUCED_80_):: defines_var(X_INTRODUCED_80_);
constraint ortools_table_int(X_INTRODUCED_83_,X_INTRODUCED_77_);
constraint array_var_int_element(X_INTRODUCED_85_,[1,3,2],3);
constraint array_int_element(X_INTRODUCED_68_,X_INTRODUCED_69_,X_INTRODUCED_86_):: defines_var(X_INTRODUCED_86_);
constraint ortools_table_int(X_INTRODUCED_89_,X_INTRODUCED_77_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_66_,X_INTRODUCED_71_],-1):: domain:: defines_var(X_INTRODUCED_71_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_67_,X_INTRODUCED_79_],-1):: domain:: defines_var(X_INTRODUCED_79_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_68_,X_INTRODUCED_85_],-1):: domain:: defines_var(X_INTRODUCED_85_);
solve  satisfy;
```

Curiously, if I change the index set of `arr_var1` to
```
array[int] of var int: arr_var1 = [T1] ++ [ rt(r) | r in 1..2 ];
```

The instance is UNSAT also in chuffed.

It's possible this is a bug in the minizinc toolchain but it's surprising that no other solver back-ends I've tried (gecode, chuffed, highs, coinbc) give UNSAT

**Comment by CervEdin** [comment]
I should note that the flatzinc for _chuffed_ was _identical_, except the name of the table predicate

**Comment by lperron** [comment]
I am debugging. Thanks

**Comment by lperron** [comment]
I reduced it to:

```
var 1..3: x ::var_is_introduced :: is_defined_var;
constraint array_var_int_element(x,[1,3,2],3);
constraint int_lin_eq([1,-1],[1 ,x],-1):: domain:: defines_var(x);
solve satisfy;
```

**Comment by CervEdin** [comment]
Seems like removing the domain consistency annotation on `int_lin_eq` makes it feasible
```
var 2..2: x ::var_is_introduced :: is_defined_var;
constraint array_var_int_element(x,[1,3,2],3);
constraint int_lin_eq([1,-1],[1 ,x],-1):: defines_var(x);
solve satisfy;
```

**Comment by CervEdin** [comment]
Looks like it's the same for the "original" flatzinc
```
predicate ortools_table_int(array [int] of var int: x,array [int] of int: t);
array [1..2] of int: X_INTRODUCED_31_ = [480,344];
array [1..12] of int: X_INTRODUCED_33_ = [480,9,7,480,9,10,480,106,7,480,106,10];
var 1..2: X_INTRODUCED_16_;
var 1..2: X_INTRODUCED_17_;
var 1..2: X_INTRODUCED_18_;
var 167..167: X_INTRODUCED_20_;
var 102..102: X_INTRODUCED_21_;
var 79..79: X_INTRODUCED_22_;
var 1..3: X_INTRODUCED_23_;
var 1..3: X_INTRODUCED_24_;
var 1..3: X_INTRODUCED_25_;
var 2..3: X_INTRODUCED_28_ ::var_is_introduced :: is_defined_var;
var 344..480: X_INTRODUCED_30_ ::var_is_introduced :: is_defined_var;
var 2..3: X_INTRODUCED_35_ ::var_is_introduced :: is_defined_var;
var 344..480: X_INTRODUCED_36_ ::var_is_introduced :: is_defined_var;
var 2..3: X_INTRODUCED_39_ ::var_is_introduced :: is_defined_var;
var 344..480: X_INTRODUCED_40_ ::var_is_introduced :: is_defined_var;
array [1..3] of var int: X_INTRODUCED_32_ ::var_is_introduced  = [X_INTRODUCED_30_,9,7];
array [1..3] of var int: X_INTRODUCED_37_ ::var_is_introduced  = [X_INTRODUCED_36_,106,10];
array [1..3] of var int: X_INTRODUCED_41_ ::var_is_introduced  = [X_INTRODUCED_40_,106,7];
constraint array_var_int_element(X_INTRODUCED_28_,[167,102,79],102);
constraint array_int_element(X_INTRODUCED_16_,X_INTRODUCED_31_,X_INTRODUCED_30_):: defines_var(X_INTRODUCED_30_);
constraint ortools_table_int(X_INTRODUCED_32_,X_INTRODUCED_33_);
constraint array_var_int_element(X_INTRODUCED_35_,[167,102,79],102);
constraint array_int_element(X_INTRODUCED_17_,X_INTRODUCED_31_,X_INTRODUCED_36_):: defines_var(X_INTRODUCED_36_);
constraint ortools_table_int(X_INTRODUCED_37_,X_INTRODUCED_33_);
constraint array_var_int_element(X_INTRODUCED_39_,[167,102,79],102);
constraint array_int_element(X_INTRODUCED_18_,X_INTRODUCED_31_,X_INTRODUCED_40_):: defines_var(X_INTRODUCED_40_);
constraint ortools_table_int(X_INTRODUCED_41_,X_INTRODUCED_33_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_16_,X_INTRODUCED_28_],-1):: defines_var(X_INTRODUCED_28_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_17_,X_INTRODUCED_35_],-1):: defines_var(X_INTRODUCED_35_);
constraint int_lin_eq([1,-1],[X_INTRODUCED_18_,X_INTRODUCED_39_],-1):: defines_var(X_INTRODUCED_39_);
solve  satisfy;
```

**Comment by lperron** [comment]
nice find. It looks to be in the only remaining flatzinc presolve code

**Comment by CervEdin** [comment]
That's great news @lperron, will this fix be available in the next release (9.12?) or will there be a new build of 9.11?

Thank you very much for the quick response and for fixing this

**Comment by lperron** [comment]
9.12.

We will start the release process next week. Hopefully, it will not take
too long.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le ven. 4 oct. 2024 à 16:38, Erik Cervin Edin ***@***.***> a
écrit :
